torchao/prototype/dtypes/uintgen.py

```diff
@@ -0,0 +1,362 @@
+            ((uint8_data[:, 3] & 0b00001111) << 4) | ((uint8_data[:, 4] & 0b00011110) >> 1),
+            ((uint8_data[:, 4] & 0b00000001) << 7) | ((uint8_data[:, 5] & 0b00011111) << 2) | ((uint8_data[:, 6] & 0b0011000) >> 3),
+            ((uint8_data[:, 6] & 0b00000111) << 5) | (uint8_data[:, 7] & 0b00011111),
+        ),
+        dim=-1,
+    ).view(down_size_uint5(shape))
+
+    return packed_data
+
+
+def unpack_uint5(packed_data: torch.Tensor) -> torch.Tensor:
+    """Unpack the 5 bytes into the 5 lowest bits of 8 bytes
+    01234567|01234567|01234567|01234567|01234567
+    AAAAABBB|BBCCCCCD|DDDDEEEE|EFFFFFGG|GGGHHHHH
+    """
+    shape = packed_data.shape
+    assert (
+        shape[-1] % 5 == 0
+    ), f"Input last dimension should be divisible by 5, but got {shape[-1]}"
+
+    packed_data = packed_data.contiguous().view(-1, 5)
+
+    unpacked_data = torch.stack(
+        (
+            ((packed_data[:, 0] >> 3) & 0b00011111),
+            ((packed_data[:, 0] & 0b00000111) << 2) | ((packed_data[:, 1] >> 6) & 0b00000011),
+            ((packed_data[:, 1] >> 1) & 0b00011111),
+            ((packed_data[:, 1] & 0b00000001) << 4) | ((packed_data[:, 2] >> 4) & 0b00001111),
+            ((packed_data[:, 2] & 0b00001111) << 1) | ((packed_data[:, 3] >> 7) & 0b00000001),
+            ((packed_data[:, 3] >> 2) & 0b00011111),
+            ((packed_data[:, 3] & 0b00000011) << 3) | ((packed_data[:, 4] >> 5) & 0b00000111),
+            packed_data[:, 4] & 0b00011111,
+        ),
+        dim=-1,
+    ).view(up_size_uint5(shape))
+
+    return unpacked_data
+
+
+def down_size_uint6(size):
+    assert size[-1] % 4 == 0, f"{size} last dim not divisible by four"
+    return (*size[:-1], size[-1] // 4 * 3)
+
+
+def up_size_uint6(size):
+    assert size[-1] % 3 == 0, f"{size} last dim not divisible by three"
+    return (*size[:-1], size[-1] // 3 * 4)
+
+
+def pack_uint6(uint8_data: torch.Tensor) -> torch.Tensor:
+    """Pack the 6 lowest bits of 4 input bytes into 3 bytes
+
+    4 -> 3
+    01234567|01234567|01234567
+    AAAAAABB|BBBBCCCC|CCDDDDDD
+
+    The packing pattern:
+    - First byte:  (A0 A1 A2 A3 A4 A5 B0 B1)
+    - Second byte: (B2 B3 B4 B5 C0 C1 C2 C3)
+    - Third byte:  (C4 C5 D0 D1 D2 D3 D4 D5)
+    """
+    shape = uint8_data.shape
+    assert (
+        shape[-1] % 4 == 0
+    ), f"Input last dimension should be divisible by 4, but got {shape[-1]}"
+
+    uint8_data = uint8_data.contiguous().view(-1, 4)
+
+    packed_data = torch.stack(
+        (
+            ((uint8_data[:, 0] & 0b00111111) << 2) | ((uint8_data[:, 1] >> 4) & 0b00000011),
+            ((uint8_data[:, 1] & 0b00001111) << 4) | ((uint8_data[:, 2] >> 2) & 0b00001111),
+            ((uint8_data[:, 2] & 0b00000011) << 6) | (uint8_data[:, 3] & 0b00111111),
+        ),
+        dim=-1,
+    ).view(down_size_uint6(shape))
+
+    return packed_data
+
+
+def unpack_uint6(packed_data: torch.Tensor) -> torch.Tensor:
```

Comment:
btw how does this wcompare to @vayuda's work?

Reply:
There is a bit of rework. After discussio on cuda-mode between @andreaskoepf  and @vayuda. Some of the odd bit functionality will be merged into @vayuda's algorithimic bitpacking.py. However will use this implementation to test correctness and speed with both approaches.